File: bernoulli-numbers/erlang/bernoulli.erl
%% Code written by Leo Vainio.

-module(bernoulli).

%% The functions that can be called from this module.
-export([bernoulli/1]).


% Calculates and prints the n first bernoulli numbers.
bernoulli(N) -> 
    B = [1],
    M = 1,

    B_nums = b_outer_loop(B, M, N),

    print_result(B_nums, 0).

% Prints out all elements of a list.
print_result([], _) -> 
    ok;
print_result([H|T], N) ->
    io:fwrite("~pth bernoulli nr: ~p\n", [N, H]),
    print_result(T, N+1).

% Calculates the outer loop in the bernoulli function.
b_outer_loop(B, M, N) when M > N -> B;
b_outer_loop(B, M, N) when M =< N ->
    K = 0,
    Bm = b_inner_loop(B, K, M, 0),
    B_new = B ++ [Bm / (M+1)],
    b_outer_loop(B_new, M+1, N).

% Calculates the inner loop in the bernoulli function.
b_inner_loop(_, K, M, Bm) when K == M -> Bm;
b_inner_loop(B, K, M, Bm) when K < M ->
    Bm_new = Bm - binom(M+1, K) * lists:nth(K+1, B),
    b_inner_loop(B, K+1, M, Bm_new).

% Calculates the binomial (n, k).
binom(N, K) ->
    R = 1,
    I = 1,
    binom_loop(R, N, K, I).

% Binomial "for loop".
binom_loop(R, _, K, I) when I > K -> R;
binom_loop(R, N, K, I) when I =< K ->
    R_new = R * (N - I + 1) / I,
    binom_loop(R_new, N, K, I + 1).

File: bernoulli-numbers/php/bernoulli.php
<?php
// Code written by Leo Vainio

// Calculate the binomial (n, k).
function binom(int $n, int $k) {
    $r = 1;
    for ($i = 1; $i <= $k; $i++) {
        $r = $r * ($n - $i + 1) / $i;
    }
    return $r;
}

// Calculate the first n bernoulli numbers.
function bernoulli(int $n) {
    $b_nums = array(1);
    for($m = 1; $m <= $n; $m++) {
        array_push($b_nums, 0);

        for($k = 0; $k < $m; $k++) {
            $b_nums[$m] = $b_nums[$m] - binom($m + 1, $k) * $b_nums[$k];
        }

        $b_nums[$m] = $b_nums[$m] / ($m + 1);
    }

    return $b_nums;
}

// Prints the n first bernoulli numbers.
function print_n_bernoulli($n) {
    $b_nums = bernoulli($n);

    for($i = 0; $i <= $n; $i++) {
        echo "$i";
        echo "th: ";
        echo $b_nums[$i];
        echo "\n";
    }
}

print_n_bernoulli(20)
?>